### DISCUSSION TITLE
Running oshi with jdk6

### DISCUSSION THREAD

**Discussion by moley** [oshi/oshi#1755] [General]
Hi, 

we'd like to use oshi to build an agent outside the datacenter to collect hardware infos of our clients. We indeed still have to support jdk6 (32bit). What would we have to do to make this happen?

Any help welcome, 
Thanks in advance

**Comment by dbwiddis** [comment]
See comments on #1687. It’s possible to backport 3.13. 

Mostly quick fixes. Hardest part is file IO. 

Since this is the third ask for this recently I may look into doing this backport myself on the next couple of weeks.

**Reply by moley** [reply]
This would be really great, thank you

**Comment by dbwiddis** [comment]
Quick update.

Good news: It was easy to restore Java 6 compatibility with the few simple patterns noted on #1687: fixing diamond operators, multi-exception catch, try-with-resources, switch/case with strings, and underscores in numbers.   The resulting commit of all those fixes is in the java6 branch here:
https://github.com/oshi/oshi/tree/java6

Bad news: I can't figure out yet how to actually build and publish that code.  Latest JDKs require a minimum Java 7 version.  
```
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] Source option 6 is no longer supported. Use 7 or later.
[ERROR] Target option 6 is no longer supported. Use 7 or later.
```
It seems one can run JDK8 to release with  1.6 compatibility but I haven't been able to figure that part out.

**Reply by mprins** [reply]
You need to downgrade some Maven plugins (at least compiler plugin) as well (and/or possibly Maven itself - which will result in more problems because Maven central no longer supports http access - it's a can of worms)

Perhaps @moley 's company can chip in with some sponsoring (having written that I see @dbwiddis has not set up sponsoring)

**Reply by dbwiddis** [reply]
> (having written that I see @dbwiddis has not set up sponsoring)

I have, actually.  :)

As far as downgrading versions, I figured as much and will experiment when I have the time.

**Comment by dbwiddis** [comment]
Quick update 2: resolved the compilation problem. Maven was using my default JDK (17) but simply setting `JAVA_HOME` to the JDK8 install worked and the code compiled.  Linking to JDK methods that don't exist is a different matter.

Good News: I have a specific list of files and lines that need fixing.

Bad News: Lots of `java.nio.*` code that needs to be ported, which I avoided ever doing in the first place by making JDK7 the minimum version a long time ago.

@moley are you aware of any libraries that backport Java 7 NIO to earlier versions?

**Comment by hazendaz** [comment]
Old guava library but its most likely vulnerable.

**Comment by dbwiddis** [comment]
I've published a Java 6 compatible (I think) version of 3.13.6 here:
https://oss.sonatype.org/content/repositories/snapshots/com/github/oshi/oshi-core-java6/3.13.6-SNAPSHOT/

@moley please test it.

**Reply by moley** [reply]
Hi @dbwiddis, 
thanks a lot. 

Looks good on the first view. 
I will go into detail testing the next days, will get feedback after done or on finding any issue

Have a nice evening meanwhile

**Comment by moley** [comment]
Hi, I have set up an multi operatingsystem test setup, and with jdk6 on centos 6.3 at least (but I a think it's a general drawback regarding jdk6) I get 
  "java.lang.NoSuchMethodError: com.sun.management.OperatingSystemMXBean.getSystemCpuLoad()D",
        "\tat oshi.hardware.common.AbstractCentralProcessor.initMXBean(AbstractCentralProcessor.java:130)",
        "\tat oshi.hardware.common.AbstractCentralProcessor.<init>(AbstractCentralProcessor.java:118)",
        "\tat oshi.hardware.platform.linux.LinuxCentralProcessor.<init>(LinuxCentralProcessor.java:57)",
        "\tat oshi.hardware.platform.linux.LinuxHardwareAbstractionLayer.getProcessor(LinuxHardwareAbstractionLayer.java:70)",

Should I open a ticket?

**Reply by dbwiddis** [reply]
> Should I open a ticket?

Nah, that's just me forgettingI used JDK7 Oracle MXBean methods.  Easily removed, I'll get another snapshot out later today.

**Reply by dbwiddis** [reply]
(And I didn't catch it on my own testing because on OpenJDK 6, it doesn't even attempt the Oracle code path.)

**Reply by dbwiddis** [reply]
New snapshot posted, same link, pick the newer JAR:  https://oss.sonatype.org/content/repositories/snapshots/com/github/oshi/oshi-core-java6/3.13.6-SNAPSHOT/

**Reply by moley** [reply]
no problem, for this have created my test setup , thanks a lot. I will retry

**Reply by moley** [reply]
I have tested on various (13 windows and linux versions) and it looks good on first sight, tomorrow I will validate the measurements themselves...  thanks a lot for fast fix

**Reply by moley** [reply]
ok, let me sum up my experiences ( I have tested it against Virtualbox VM as said, perhaps this disturbs my validation?)
* I try to get the load of the processor via systemInfo.getHardware().getProcessor().getSystemLoadAverage(1 / 2 / 3) which works good on linux systems, but returns -1 for windows. Is there any alternative to get the requested data, is this a known issue or do I have to check it with real hardware? 
* On linux systems I try to get total and free mem via systemInfo.getHardware().getComputerSystem().getAvailable() and getTotal(). While total mem is accurate, the free mem seems to be not. Free tells me I have 377916k of free memory while oshi tells me to have 444325888 (=433912k) while I am running the agent, which uses oshi. I would have expected to be less than free tells me. Could you imagine what makes this difference?

Would be nice hearing from you.

**Reply by dbwiddis** [reply]
Windows does not have load average. The -1 return is documented in the Javadoc for that method. 

The memory issue has to do with the definition of “free”.  OSHI reports “Available” which includes some portion of cache/buffers.  “Free” always drops to near zero on Linux and is useless as a stat.  We pull directly from `/proc/meminfo` for MemAvailable which is the official place to get that statistic.

**Reply by dbwiddis** [reply]
Based on this review I'm going to merge this branch into the regular 3.X branch.  I'll spend a few days looking over any bug fixes in the 4.X and 5.X series that should be backported.

**Reply by moley** [reply]
Ok, thank you for backporting..... 
The memory thing is OK for us I guess - I'd only liked to understand - Will compare it against /proc/meminfo in tests. 
The thing with the load average on windows.... I did not yet read the javadoc I must admit, but I had a suspicion that this is not available. We'd like to have an information how heavy the CPU is on fire to decide if we are able to add additional parallel programs on the machines. So a measurement if the processor is already on fire would be fine. Can you imagine an alternative way of getting this?

**Reply by dbwiddis** [reply]
@moley You may find this discussion helpful in understanding why we choose MemAvailable: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=34e431b0ae398fc54ea69ff85ec700722c9da773

As far as CPU load, OSHI reports the "tick counts" at any given moment, either on a systemwide or per-processor basis.  To measure CPU load, you can just collect the ticks at two different points in time and then calculate the difference in ticks (load) divided by the difference in time.  

In theory if you collected ticks at intervals close to 1/5/15 minutes you could replicate the Cpu Load statistic, but you'd have to save up a lot of ticks to enable that math.   There is a `getSystemCpuLoadBetweenTicks()` method (and also `getProcessorCpuLoadBetweenTicks()`) in OSHI 3.X that automatically does this math for you, measuring between "the last time you called the method".  Note the API for those methods changed in 4.X.

**Comment by dbwiddis** [comment]
Separate subject. When I started the port I thought I'd make a separate java6 artifact, but in reality there's really no different functionality and I'm contemplating just incrementing the 3.X branch with this release.  Maybe make it 3.14 (or 3.14.15926....
